### ISSUE TITLE
Index element properties with mutable data

### ISSUE THREAD

**Issue by jacomyal** [medialab/graphlib#4]
I think keeping indexes of nodes' and edges' properties up to date might be incompatible with mutable nodes and edges. An efficient events system that can trigger on nodes and edges updates can keep indexes updated, but to do so nodes and edges must be somehow observable.

One solution would be to have immutable (even frozen, maybe?) nodes and edges, and methods to update some nodes or edges.

**Comment by DavidBruant** [comment]
I've had this question (immutable graphs) in my mind for some time :-)
I think that if we were to create an API, we should consider making it possible for both mutable and immutable graphs to implement this API.

To answer your question specifically, the graph could also be a [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) and resolve indices lazily (but I prefer the immutable version by a long shot)

**Comment by Yomguithereal** [comment]
Immutability is not very an implementation problem. We could easily switch from mutable to immutable by using `Graph` or `ImmutableGraph` even if the latter would mean to freeze some internal objects and a persistent data structure.

**Comment by DavidBruant** [comment]
> We could easily switch from mutable to immutable 

We have to be considerate about this. For instance, for an immutable graph, the `addNode` method would need to return the new graph (it can return `undefined` if the graph is mutable and I subject it does anytime the graph is mutable)

**Comment by Yomguithereal** [comment]
Yes, I guess adding a node with an immutable graph should return a new graph indeed. This could be done efficiently with an underlying persistent data structure (which is very easy to implement in our case since we just have to shallow clone two arrays).

**Comment by Yomguithereal** [comment]
We could alternatively design a "real" persistent data structure to write in `O(nlogn)` but I am not sure this is within the scope of this lib since it would mean a whole lot of code.

**Comment by DavidBruant** [comment]
> We could alternatively design a "real" persistent data structure to write in O(log n) but I am not sure this is within the scope of this lib since it would mean a whole lot of code.

I we were to create an immutable graph, I'd just wrap two [immutablejs Sets](https://facebook.github.io/immutable-js/docs/#/Set) (one for nodes, one for edges) inside our API and call it a day (more or less). Persistent data structure is its own work indeed, let's sit on shoulders of other giants.

**Comment by Yomguithereal** [comment]
Good point. But you'll agree we would need to require Immutable Graphs differently from Graphs since it would mean to import quite a huge dependency (Immutable.js) is far from being a light library.

**Comment by DavidBruant** [comment]
Sure. An immutable graph would only be a single implementation of the agreed upon API.

> since it would mean to import quite a huge dependency (Immutable.js) is far from being a light library.

I'm getting off topic, but with ES6 modules and things like http://rollupjs.org I wouldn't worry too much about "large" libraries; they'll soon (one year? two?) be stripped down to just what you need automatically as part of the build process.

**Comment by Yomguithereal** [comment]
Fair enough. We can also rely on loose peer dependencies, like, if you want it to work, think to install Immutable.js but we won't do it for you.
